packages/grafana-ui/src/components/Combobox/useOptions.ts

```diff
@@ -0,0 +1,82 @@
+import { debounce } from 'lodash';
+import { useState, useCallback, useMemo } from 'react';
+
+import { itemFilter } from './filter';
+import { ComboboxOption } from './types';
+import { StaleResultError, useLatestAsyncCall } from './useLatestAsyncCall';
+
+type AsyncOptions<T extends string | number> =
+  | Array<ComboboxOption<T>>
+  | ((inputValue: string) => Promise<Array<ComboboxOption<T>>>);
+
+const asyncNoop = () => Promise.resolve([]);
+
+/**
+ * Abstracts away sync/async options for MultiCombobox (and later Combobox).
+ * It also filters options based on the user's input.
+ *
+ * Returns:
+ *  - options either filtered by user's input, or from async options fn
+ *  - function to call when user types (to filter, or call async fn)
+ *  - loading and error states
+ */
+export function useOptions<T extends string | number>(rawOptions: AsyncOptions<T>) {
+  const isAsync = typeof rawOptions === 'function';
+
+  const loadOptions = useLatestAsyncCall(isAsync ? rawOptions : asyncNoop);
```

Comment:
What if instead of `asyncNoop` we returned here:
```
(input) => rawOptions.filter(itemFilter(input))
```
I feel like that way we don't have to use `isAsync` switch anywhere.

The debouncing may be weird, but then maybe not and it would be ok that even local filter does not jump around with every character input. Or the debounce time can be 0 in that case.

Not super important but thought it could make this a bit cleaner.

Reply:
Hmmmmmm. Interesting idea - this is what you were getting at in the earlier call.

I like it in theory, but two potential problems that would need working with:
 - debounced filtering like you mention
 - loading states used unnecessarily

I'll give it a try and see how it goes.

Reply:
Okay - I took a stab at this and... I think it could work

https://github.com/grafana/grafana/compare/joshhunt/multicombobox-async...joshhunt/simpler-use-options-fn#diff-fa1efd5993aebed9bd81a3430a5a453c786ca5378276fc74dde89ab6d7ee769dR24

Loading state ends up being a no-op because `isLoading(true)` and `isLoading(false)` is set in the same tick, so ends up being a no op I think. 

Debounced sync is... fine I think? It is different, but I don't feel noticeably wrong.

One interesting side effect  (advantage) is that it allows for `options: (term) => ComboboxOption[]` API which would allow consumers to implement their own filtering without opting into async.